File: Search and Sort/MergeSort.cpp
#include <cstdio>
#include <cstring>
using namespace std;

void mergeArray(int array[], int left, int middle, int right) {
    int sorted[right - left + 1];

    int cur = 0;
    int s1  = left;
    int s2  = middle + 1;
    while(s1 <= middle && s2 <= right) {
        if (array[s1] <= array[s2]) {
            sorted[cur++] = array[s1++];
        } else {
            sorted[cur++] = array[s2++];
        }
    }

    while(s1 <= middle) {
        sorted[cur++] = array[s1++];
    }

    while(s2 <= right) {
        sorted[cur++] = array[s2++];
    }

    memcpy(&array[left], sorted, sizeof(int) * (right - left + 1));
}

void mergeSort(int array[], int left, int right) {
    if (left < right) {
        int middle = left + (right - left) / 2;
        mergeSort(array, left, middle);
        mergeSort(array, middle + 1, right);
        mergeArray(array, left, middle, right);
    }
}

int printArray(int array[], int length) {
    for (int i = 0; i < length; i++) {
        printf("%d ", array[i]);
    }
}

int main(int argc, char **argv) {
    int array[] = { 12, 11, 13, 5, 6, 7 };
    int length = sizeof(array) / sizeof(int);
    mergeSort(array, 0, length - 1);
    printArray(array, length);
}

File: Search and Sort/MergeSort.py
def mergeArray(array, left, middle, right):
    sorted = [0] * (right - left + 1)

    cur = 0
    s1  = left
    s2  = middle + 1
    while (s1 <= middle) and (s2 <= right):
        if (array[s1] <= array[s2]):
            sorted[cur] = array[s1]
            s1 = s1 + 1
        else:
            sorted[cur] = array[s2]
            s2 = s2 + 1
        cur = cur + 1

    while s1 <= middle:
        sorted[cur] = array[s1]
        s1 = s1 + 1
        cur = cur + 1

    while s2 <= right:
        sorted[cur] = array[s2]
        s2 = s2 + 1
        cur = cur + 1

    for cur in range(left, right + 1):
        array[cur] = sorted[cur - left]

def mergeSort(array, left, right):
    if (left < right):
        middle = left + int((right - left) / 2)
        mergeSort(array, left, middle)
        mergeSort(array, middle + 1, right)
        mergeArray(array, left, middle, right)

if __name__ == "__main__":
    array = [ 12, 11, 13, 5, 6, 7 ]
    mergeSort(array, 0, len(array) - 1)
    print(array)